File: problems/queue_machine/submissions/accepted/queue_machine.cpp
#include <iostream>
#include <string>
#include <deque> 
#include <ctype.h>

using namespace std;

void showdq(deque <int> g,string s,deque <int> :: iterator it) 
{	
	if(g.size()){
		if(s.compare("ALL")==0){ 
    		for (it = g.begin(); it != g.end(); ++it) 
        		cout << *it << " "; 
			cout << endl;
		}else if(s.compare("FIRST")==0){
			cout << g.at(0)<< endl;
		}
    	else
		{
			cout << *g.rbegin()<<endl;
		}
	}else
	{
		cout << 0 <<endl;
	}	 
}

int main()
{
	int ticket_number = 0;
	int n_of_commands;
	cin >> n_of_commands;

	string commands[n_of_commands];
	for(int i=0;i<n_of_commands;i++){
		cin >> commands[i];
		if(commands[i].compare("CANCEL")==0 || commands[i].compare("SHOW")==0)
			i--;
	}

	deque <int>queue;
	deque <int> :: iterator it;

	for(int i=0;i<n_of_commands;i++){
		if(commands[i].compare("NEW")==0){
			ticket_number++;
			queue.push_back(ticket_number);
		}
		else if(!isalpha(commands[i][0])){
			it =queue.begin()+stoi(commands[i]);
			queue.erase(it); 
		}
		else if(commands[i].compare("NEXT")==0){
			if(queue.size())
				queue.pop_front();
		}
		else{
			showdq(queue,commands[i],it);
		}		
	}

	return 0;
}

File: problems/queue_machine/submissions/accepted/queue_machine.go
package main

import (
	"fmt"
	"strconv"
)

func find(queue []int, v int) int {
	for i, n := range queue {
		if v == n {
			return i
		}
	}
	return len(queue)
}

func main() {
	var cmdNumber int
	fmt.Scanf("%d", &cmdNumber)
	queue := make([]int, 0)
	ticketNumber := 0

	var cmd, arg string
	for i := 0; i < cmdNumber; i++ {
		fmt.Scanln(&cmd, &arg)
		switch cmd {
		case "NEW":
			ticketNumber++
			queue = append(queue, ticketNumber)
		case "CANCEL":
			orderNumber, _ := strconv.Atoi(arg)
			orderindex := find(queue, orderNumber)
			queue = append(queue[:orderindex], queue[orderindex+1:]...)
		case "SHOW":
			switch arg {
			case "FIRST":
				fmt.Println(queue[0])
			case "LAST":
				fmt.Println(queue[len(queue)-1])
			case "ALL":
				if len(queue) == 0 {
					fmt.Println(0)
				} else {
					for _, v := range queue[:len(queue)-1] {
						fmt.Printf("%d ", v)
					}
					fmt.Println(queue[len(queue)-1])
				}
			}
		case "NEXT":
			queue = queue[1:]
		}
	}
}
